e2e_playwright/websocket_reconnects_test.py

```diff
@@ -0,0 +1,189 @@
+    """Wait for the expected_status to appear in the status widget.
+
+    Uses the browser's MutationObserver API to observe changes to the DOM. This way,
+    we will never have a race condition between calling disconnect and checking the
+    status.
+    If the status is not observed within 1 second, the promise will resolved with an
+    error message. We don't use reject because on Firefox this seem to cause an
+    undefined error which is not as precise as our error message.
+    Otherwise, the promise is resolved with the status.
+    """
+
+    return app.evaluate(
+        """async ([expectedStatus]) => {
+                const p = new Promise((resolve) => {
+                    // Define a timeoutId so that we can cancel the timeout in the
+                    // callback upon success
+                    let timeoutId = null
+
+                    const callback = (mutationList, observer) => {
+                        for (const mutation of mutationList) {
+                            console.log(mutation)
+                            if (mutation.type !== "childList") {
+                                continue
+                            }
+                            if (mutation.addedNodes.length === 0) {
+                                continue
+                            }
+                            for (const node of mutation.addedNodes) {
+                                const testId = node.getAttribute('data-testid')
+                                if (testId === 'stStatusWidget') {
+                                    const status = node.textContent
+                                    if (status.indexOf(expectedStatus) > -1) {
+                                        if (timeoutId) clearTimeout(timeoutId)
+                                        if (observer) observer.disconnect()
+                                        resolve(status)
+                                        return
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    const observer = new MutationObserver(callback);
+                    // Observe toolbar for changes, which includes status widget
+                    const targetNode = document.querySelector('[data-testid=stToolbar]')
+                    if (!targetNode) {
+                        resolve("toolbar not found")
+                        return
+                    }
+                    const config = { childList: true, subtree: true };
+                    observer.observe(targetNode, config);
+
+            """
+        + callable_action
+        + """
+                    timeoutId = setTimeout(() => {
+                        if (observer) observer.disconnect()
+                        resolve(`timeout: did not observe status '${expectedStatus}'`)
+                        return
+                    }, 3000);
+                })
```

Comment:
I don't think this is possible in practice, but should `callable_action` be called after the timeout is set so that the scenario where the callback files and clears the timeout before `timeoutId` is actually set is truly impossible?

Reply:
good point. I chose this sequence in case the passed callable action itself already takes some time to execute (which is desired) and not taking any time from the timeout-budget. I think the scenario you described could happen, however: calling `clearTimeout` on `undefined` is a no-op, and even if the timeout runs, calling `observer.disconnect` multiple times is also a no-op, and calling `resolve` after it has been called once is also a no-op. I have left a comment for this, but would not introduce any additional flags etc. to keep the code concise (since it is inline in Python, its already not that nice to read).
I hope this makes sense 🙂